File: server/billing/ports.ts
import type { VaultContext } from '../../lib/domain/identity';
import type { BillingSubscriptionRecord } from './core';
import type { ProviderSubscriptionCancellationCommand } from './cancellation-core';
import type {
  CheckoutIntentRecord,
  ProviderEventReceipt,
  ReconciliationCheckpoint,
} from './records';
import type {
  BillingProvider,
  BillingOwnerScope,
  BillingSubscriptionId,
  CheckoutIntentId,
  ProviderCustomerReference,
  ProviderCheckoutReference,
  ProviderEventId,
  ProviderSubscriptionReference,
  ReconciliationSnapshotId,
} from './public';

export type BillingOwnershipPort = {
  owns(context: VaultContext): Promise<boolean>;
};

export type CheckoutCreateResult =
  | { readonly kind: 'created' }
  | {
      readonly kind: 'existing';
      readonly record: BillingSubscriptionRecord | undefined;
      readonly intent: CheckoutIntentRecord | undefined;
    };

export type CheckoutOpenResult =
  | { readonly kind: 'applied' }
  | { readonly kind: 'replayed' }
  | { readonly kind: 'conflict' };

export type AtomicCommitResult =
  | { readonly kind: 'applied' }
  | { readonly kind: 'duplicate' }
  | { readonly kind: 'conflict' };

export type ReceiptRecordResult =
  | { readonly kind: 'recorded' }
  | { readonly kind: 'duplicate' };

export type BillingRepository = {
  findByOwner(
    context: BillingOwnerScope,
  ): Promise<BillingSubscriptionRecord | undefined>;
  findById(
    subscriptionId: BillingSubscriptionId,
  ): Promise<BillingSubscriptionRecord | undefined>;
  findByProviderMapping(input: {
    readonly provider: BillingProvider;
    readonly customerReference: ProviderCustomerReference;
    readonly subscriptionReference: ProviderSubscriptionReference;
  }): Promise<BillingSubscriptionRecord | undefined>;
  findCheckoutIntent(
    checkoutIntentId: CheckoutIntentId,
  ): Promise<CheckoutIntentRecord | undefined>;
  findCheckoutByProviderReference(
    provider: BillingProvider,
    reference: ProviderCheckoutReference,
  ): Promise<CheckoutIntentRecord | undefined>;
  createCheckout(input: {
    readonly record: BillingSubscriptionRecord;
    readonly intent: CheckoutIntentRecord;
  }): Promise<CheckoutCreateResult>;
  openCheckout(input: {
    readonly context: VaultContext;
    readonly intent: CheckoutIntentRecord;
  }): Promise<CheckoutOpenResult>;
  findProviderEventReceipt(
    provider: BillingProvider,
    eventId: ProviderEventId,
  ): Promise<ProviderEventReceipt | undefined>;
  commitProviderFact(input: {
    readonly expectedRecord: BillingSubscriptionRecord;
    readonly nextRecord: BillingSubscriptionRecord;
    readonly receipt: ProviderEventReceipt;
  }): Promise<AtomicCommitResult>;
  recordIgnoredProviderFact(
    receipt: ProviderEventReceipt,
  ): Promise<ReceiptRecordResult>;
  findReconciliationCheckpoint(
    provider: BillingProvider,
    snapshotId: ReconciliationSnapshotId,
  ): Promise<ReconciliationCheckpoint | undefined>;
  commitReconciliation(input: {
    readonly expectedRecord: BillingSubscriptionRecord;
    readonly nextRecord: BillingSubscriptionRecord;
    readonly checkpoint: ReconciliationCheckpoint;
  }): Promise<AtomicCommitResult>;
  recordIgnoredReconciliation(
    checkpoint: ReconciliationCheckpoint,
  ): Promise<ReceiptRecordResult>;
};

export type SubscriptionCancellationProviderPort = {
  cancelSubscription(
    command: ProviderSubscriptionCancellationCommand,
  ): Promise<unknown>;
};

File: backend/internal/billing/ports.go
package billing

import (
	"context"

	"github.com/fukamu/notes/backend/internal/identity"
)

type OwnershipPort interface {
	Owns(context.Context, identity.VaultContext) (bool, error)
}

type CheckoutCreateKind string

const (
	CheckoutCreated  CheckoutCreateKind = "created"
	CheckoutExisting CheckoutCreateKind = "existing"
)

type CheckoutCreateResult struct {
	Kind   CheckoutCreateKind
	Record *SubscriptionRecord
	Intent *CheckoutIntentRecord
}

type CommitKind string

const (
	CommitApplied   CommitKind = "applied"
	CommitReplayed  CommitKind = "replayed"
	CommitDuplicate CommitKind = "duplicate"
	CommitConflict  CommitKind = "conflict"
)

type BillingRepository interface {
	FindByOwner(context.Context, OwnerScope) (*SubscriptionRecord, error)
	FindByID(context.Context, SubscriptionID) (*SubscriptionRecord, error)
	FindByProviderMapping(
		context.Context,
		Provider,
		ProviderCustomerReference,
		ProviderSubscriptionReference,
	) (*SubscriptionRecord, error)
	FindCheckoutIntent(context.Context, CheckoutIntentID) (*CheckoutIntentRecord, error)
	FindCheckoutByProviderReference(context.Context, Provider, ProviderCheckoutReference) (*CheckoutIntentRecord, error)
	CreateCheckout(context.Context, SubscriptionRecord, CheckoutIntentRecord) (CheckoutCreateResult, error)
	OpenCheckout(context.Context, OwnerScope, CheckoutIntentRecord) (CommitKind, error)
	FindProviderEventReceipt(context.Context, Provider, ProviderEventID) (*ProviderEventReceipt, error)
	CommitProviderFact(
		context.Context,
		SubscriptionRecord,
		SubscriptionRecord,
		ProviderEventReceipt,
	) (CommitKind, error)
	RecordIgnoredProviderFact(context.Context, ProviderEventReceipt) (CommitKind, error)
	FindReconciliationCheckpoint(
		context.Context,
		Provider,
		ReconciliationSnapshotID,
	) (*ReconciliationCheckpoint, error)
	CommitReconciliation(
		context.Context,
		SubscriptionRecord,
		SubscriptionRecord,
		ReconciliationCheckpoint,
	) (CommitKind, error)
	RecordIgnoredReconciliation(context.Context, ReconciliationCheckpoint) (CommitKind, error)
}
